_posts/2021-05-18-managing-react-application-state-with-context.md

```diff
@@ -0,0 +1,177 @@
+        ...state,
+        count: state.count - 1,
+      };
+    case "setValue": {
+      return {
+        ...state,
+        count: action.payload.value,
+      };
+    }
+    default:
+      throw Error(`Unhandled action type: ${action!.type}`);
+  }
+}
+```
+
+Next, I set up Context for both the `State` and `Dispatch` and provide a function component that determines how the `Reducer` we created above will be used. I also provide a way of setting the initial state which aids in testability since you can establish a given state in your tests before performing actions to see how the system behaves.
+
+```javascript
+// countProvider.tsx
+const CountStateContext = createContext<State | undefined>(undefined);
+const CountDispatchContext = createContext<Dispatch | undefined>(undefined);
+
+const CountProvider: FC<{ init?: Partial<State> }> = ({ children, init }) => {
+  const initialState: State = {
+    count: 0,
+    ...init,
+  };
+
+  const [state, dispatch] = useReducer(CountReducer, initialState);
+
+  return (
+    <CountStateContext.Provider value={state}>
+      <CountDispatchContext.Provider value={dispatch}>
+        {children}
+      </CountDispatchContext.Provider>
+    </CountStateContext.Provider>
+  );
+};
+```
+
+I then wrap our App in the `CountProvider` component, ensuring that all of the components that rely on the `State` or will perform `Action`s on that state are encapsulated as children of this component.
+
+```javascript
+// App.tsx
+function App() {
+  return (
+    <CountProvider>
+      <CountDisplay />
+      <CountButtons />
+      <CountInput />
+    </CountProvider>
+  );
+}
+```
+
+Finally, I provide some convenience functions which allow us to easily access the our `State` and `Dispatch` actions within components along with providing some error feedback if we are trying to access them outside of our `CountProvider`.
+
+```javascript
+// countProvider.tsx
+export function useCount(): [State, Dispatch] {
+  const state = useContext(CountStateContext);
+  const dispatch = useContext(CountDispatchContext);
+  if (state === undefined || dispatch === undefined) {
+    throw new Error("useCount must be used within a CountProvider");
+  }
+  return [state, dispatch];
+}
+
+export function increment(dispatch: Dispatch) {
+  dispatch({ type: "increment" })
+}
+
+export function decrement(dispatch: Dispatch) {
+  dispatch({ type: "decrement" })
+}
+
+export function setValue(dispatch: Dispatch, value: number) {
+  dispatch({ type: "setValue", payload: { value } });
+}
+```
+
+I've included some examples below of what this looks like within our components.
+
+```javascript
+// CountDisplay.tsx
+const CountDisplay = () => {
+  const [{ count }] = useCount();
+
+  return <div>{count}</div>;
+};
+
+// CountButtons.tsx
+const CountButtons = () => {
+  const [, dispatch] = useCount();
+
+  const onIncrement = () => {
+    increment(dispatch);
+  };
+
+  const onDecrement = () => {
+    decrement(dispatch);
+  };
+
+  return (
```

Comment:
😆 yeah this example! (as per my previous comment)

Reply:
Is this regarding the performance and `memo`? I've included the example of `useMemo` and `useCallback` in the repo, I just didn't want to make the post too long. Could be a follow up post?

Reply:
Yeah I appreciate your remake about keeping the post small; should be fine. I'll be going in depth about React Component lifecycle in a future FED Talk episode anyway. 👍